File: Scripts/Python/make-torus.py
# Use the current selection to create a toroidal universe.
# Author: Andrew Trevorrow (andrew@trevorrow.com), Sept 2010.

from glife import inside, outside
import golly as g

selrect = g.getselrect()
if len(selrect) == 0: g.exit("There is no selection.")
x =  selrect[0]
y =  selrect[1]
wd = selrect[2]
ht = selrect[3]

selcells = g.getcells(selrect)
if not g.empty():
    g.clear(inside)
    g.clear(outside)

# get current rule, remove any existing suffix, then add new suffix
rule = g.getrule().split(":")[0]
g.setrule(rule + ":T" + str(wd) + "," + str(ht))

newx = -int(wd/2)
newy = -int(ht/2)
selrect[0] = newx
selrect[1] = newy
g.select(selrect)
if len(selcells) > 0: g.putcells(selcells, newx - x, newy - y)
g.fitsel()

File: Scripts/Lua/make-torus.lua
-- Use the current selection to create a toroidal universe.
-- Author: Andrew Trevorrow (andrew@trevorrow.com), Apr 2016.

local g = golly()

local selrect = g.getselrect()
if #selrect == 0 then g.exit("There is no selection.") end

local x, y, wd, ht = table.unpack(selrect)
local selcells = g.getcells(selrect)

if not g.empty() then
    g.clear(0)
    g.clear(1)
end

-- get current rule, remove any existing suffix, then add new suffix
local rule = g.getrule()
rule = rule:match("^(.+):") or rule
g.setrule(string.format("%s:T%d,%d", rule, wd, ht))

local newx = -math.floor(wd/2)
local newy = -math.floor(ht/2)
selrect[1] = newx
selrect[2] = newy
g.select(selrect)
if #selcells > 0 then g.putcells(selcells, newx - x, newy - y) end
g.fitsel()
